solidity/contracts/bridge/Bridge.sol

```diff
@@ -801,4 +941,473 @@ contract Bridge is Ownable {
+
+    // TODO: Documentation.
+    function requestRedemption(
+        bytes20 walletPubKeyHash,
+        BitcoinTx.UTXO calldata mainUtxo,
+        bytes calldata redeemerOutputHash,
+        uint64 amount
+    ) external {
+        require(!faultyWallets[walletPubKeyHash], "Wallet marked as faulty");
+
+        bytes32 mainUtxoHash = mainUtxos[walletPubKeyHash];
+        require(mainUtxoHash != bytes32(0), "No main UTXO for given wallet");
+        require(
+            keccak256(
+                abi.encodePacked(
+                    mainUtxo.txHash,
+                    mainUtxo.txOutputIndex,
+                    mainUtxo.txOutputValue
+                )
+            ) == mainUtxoHash,
+            "Invalid main UTXO data"
+        );
+
+        // TODO: Check if `walletPubKeyHash` should be validated in some other ways.
+
+        // Validate if redeemer output hash has a correct length corresponding
+        // to a valid Bitcoin hash. P2PKH, P2WPKH and P2SH outputs will
+        // have 20-byte hashes. P2WSH outputs will have 32-byte hash.
+        require(
+            redeemerOutputHash.length == 20 || redeemerOutputHash.length == 32,
+            "Incorrect redeemer output hash length"
+        );
+
+        require(
+            keccak256(abi.encodePacked(walletPubKeyHash)) !=
+                keccak256(redeemerOutputHash),
+            "Redeemer output hash must not be equal to the wallet PKH"
+        );
+
+        require(
+            amount >= redemptionDustThreshold,
+            "Redemption amount too small"
+        );
+
+        // The redemption key is built on top of wallet public key hash
+        // and redeemer output hash pair. That means there can be only one
+        // request asking for redemption from specific wallet to the given
+        // BTC hash in the same time.
+        uint256 redemptionKey =
+            uint256(
+                keccak256(
+                    abi.encodePacked(walletPubKeyHash, redeemerOutputHash)
+                )
+            );
+
+        require(
+            // slither-disable-next-line incorrect-equality
+            pendingRedemptionRequests[redemptionKey].requestedAt == 0,
+            "Pending request with same redemption key already exists"
+        );
+
+        uint64 treasuryFee = redemptionTreasuryFee;
+        uint64 txMaxFee = redemptionTxMaxFee;
+
+        // The wallet's BTC balance should allow to cover the actual BTC amount
+        // transferred to the redeemer along with its share of the transaction
+        // fee. Those two components always sum up to the request's maximum
+        // redeemable amount which is computed as the difference between the
+        // requested amount and the treasury fee. This is why one should
+        // add the redeemable amount to the total requested value for the given
+        // wallet and assert this total value doesn't exceed the current main
+        // UTXO value  which is actually the same as the current wallet balance.
```

Comment:
I was a bit lost after reading about the transaction fee and that they together make the total redeemable amount. It all makes sense but I think it is not relevant in this context. I would rather explain the next few lines as follows:

```suggestion
        // The main wallet UTXO's value does include all pending redemptions.
        // To determine if the requested redemption can be performed by the
        // wallet we need to subtract the total value of all pending redemptions
        // from that wallet's main UTXO value. Given that the treasury fee is
        // not redeemed from the wallet, we are subtracting it.
```

Reply:
See https://github.com/keep-network/tbtc-v2/pull/128/commits/bdcf98ca90b17450b16cfe59d235ce8c119b611b